Move node related methods to framework/node package

jiatongw opened the pull request:
**What type of PR is this?**

/kind cleanup
/priority backlog

**What this PR does / why we need it**:

- Move `util.go` node related functions to `framework/node` package, alias `e2enode`
- Fix references of some methods to use the right package

Reference issue #77197, #76206 (parent issue #75601 )

**Special notes for your reviewer**:

There are a lot of node related functions in `util.go` and `service_util.go` are calling `Failf()` or `ExpectNoError()`. I leave those unchanged because moving them to `framework/node` will cause a `import cycle` error (i.e. `import e2enode` in `util.go` and `import framework` in `framework/node` package ). I plan to do follow up PRs to completely clean up the rest of node related functions.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

Commit:
Move node related methods to framework/node package

- Add a package "node" under e2e/framework and alias e2enode;
- Rename some functions whose name have redundant string.

Signed-off-by: Jiatong Wang <wangjiatong@vmware.com>

jiatongw commented:
/assign @andrewsykim @timothysc

jiatongw commented on test/e2e/framework/util.go:
```diff
@@ -135,8 +135,6 @@ const (
 	PollShortTimeout = 1 * time.Minute
-	// PollLongTimeout is the long timeout value in polling.
-	PollLongTimeout = 5 * time.Minute
 
```
This const is not in use so removed

figo commented:
/ok-to-test

yastij commented on test/e2e/apps/network_partition.go:
```diff
@@ -30,14 +30,15 @@ import (
 	"k8s.io/client-go/tools/cache"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
this change isn't needed (same below)

jiatongw replied:
This was added automatically be `goreturns` when saving files.. will remove it.

jiatongw replied:
Fixed

jiatongw commented:
/retest

jiatongw commented:
/retest

jiatongw commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,318 @@
+
+// FirstNodeAddresses returns the first address of the given type of each node.
+func FirstNodeAddresses(nodelist *v1.NodeList, addrType v1.NodeAddressType) []string {
+	hosts := []string{}
```
The original method named `NodeAddresses`. Rename this method to `FirstNodeAddresses` is because `e2enode.NodeAddresses` will cause a golint error (stutters).

andrewsykim replied:
Does this function need to return `[]string`, wondering why not just `string`? I see it was always like this but might be worth doing a refactor now, or at least leave a TODO for later

jiatongw commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,318 @@
+
+// ProxyRequest performs a get on a node proxy endpoint given the nodename and rest client.
+func ProxyRequest(c clientset.Interface, node, endpoint string, port int) (restclient.Result, error) {
+	// proxy tends to hang in some cases when Node is not ready. Add an artificial timeout for this call.
```
The original method named `NodeProxyRequest`. Rename this method to `ProxyRequest` is because `e2enode.NodeProxyRequest` will cause a golint error (stutters).

jiatongw commented:
/test pull-kubernetes-e2e-gce-storage-slow

andrewsykim commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,318 @@
+	return hosts
+}
+
+func isNodeConditionSetAsExpected(node *v1.Node, conditionType v1.NodeConditionType, wantTrue, silent bool) bool {
```
this function 🙈

jiatongw replied:
I will add a `TODO` for `FirstNodeAddresses `, and will make the change along with others in the following PR.

andrewsykim commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,318 @@
+
+const (
+	// Poll is how often to Poll pods, nodes and claims.
+	Poll = 2 * time.Second
```
can this be unexported?

jiatongw replied:
https://cs.k8s.io/?q=framework.Poll&i=nope&files=&repos=kubernetes/kubernetes

This are used in other packages.

andrewsykim replied:
That's a different constant that the one here though?

jiatongw replied:
That's the same. The `Poll` here is which I copied from `util.go`, to avoid import cycles.
But we can make this unexported, since other packages import from `framework.Poll`, we can make this `Poll` just use within node package?

andrewsykim commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,318 @@
+	Poll = 2 * time.Second
+
+	// PollShortTimeout is the short timeout value in polling.
+	PollShortTimeout = 1 * time.Minute
```
can this be unexported?

jiatongw replied:
https://cs.k8s.io/?q=PollShortTimeout&i=nope&files=&repos=kubernetes/kubernetes

This is also referred by other packages

andrewsykim commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,318 @@
+	// SingleCallTimeout is how long to try single API calls (like 'get' or 'list'). Used to prevent
+	// transient failures from failing tests.
+	// TODO: client should not apply this timeout to Watch calls. Increased from 30s until that is fixed.
+	SingleCallTimeout = 5 * time.Minute
```
can this be unexported?

jiatongw replied:
https://cs.k8s.io/?q=SingleCallTimeout&i=nope&files=&repos=kubernetes/kubernetes

same above

andrewsykim reviewed (commented):
Minor comments, overall PR looks good. I think `test/e2e/framework/node` can use some more cleaning up, but we can do this in a follow-up PR once things are moved into it's own package.

jiatongw commented:
/retest

jiatongw commented:
/retest

jiatongw commented:
/assign @oomichi

andrewsykim commented on test/e2e/framework/kubelet_stats.go:
```diff
@@ -499,7 +500,7 @@ type usageDataPerContainer struct {
 // GetKubeletHeapStats returns stats of kubelet heap.
 func GetKubeletHeapStats(c clientset.Interface, nodeName string) (string, error) {
-	client, err := NodeProxyRequest(c, nodeName, "debug/pprof/heap", ports.KubeletPort)
+	client, err := e2enode.ProxyRequest(c, nodeName, "debug/pprof/heap", ports.KubeletPort)
```
cc @alejandrox1

andrewsykim commented on test/e2e/scheduling/equivalence_cache_predicates.go:
```diff
@@ -28,11 +28,13 @@ import (
 
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
+
```
remove extra line

andrewsykim commented on test/e2e/scheduling/preemption.go:
```diff
@@ -35,10 +35,12 @@ import (
 
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
+
```
same here

andrewsykim reviewed (commented):
/lgtm

Minor comments, let's hold til after code freeze

andrewsykim commented:
/assign @alejandrox1 

For extra 👀

figo commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,316 @@
+
+// FirstNodeAddresses returns the first address of the given type of each node.
+// TODO: Use return type string instead of []string
+func FirstNodeAddresses(nodelist *v1.NodeList, addrType v1.NodeAddressType) []string {
```
could not find the original function `NodeAddresses` been removed?  is it `GetNodeAddresses`

jiatongw replied:
https://cs.k8s.io/?q=GetNodeAddresses&i=nope&files=&repos=kubernetes/kubernetes

It's here, in `test/e2e/framework/service_util.go`

jiatongw commented:
/test pull-kubernetes-node-e2e-containerd

andrewsykim commented:
/lgtm

andrewsykim commented:
/milestone v1.16

timothysc commented on test/e2e/framework/networking_utils.go:
```diff
@@ -579,7 +580,7 @@ func (config *NetworkingTestConfig) setup(selector map[string]string) {
 	ExpectNoError(WaitForAllNodesSchedulable(config.f.ClientSet, 10*time.Minute))
 	nodeList := GetReadySchedulableNodesOrDie(config.f.ClientSet)
-	config.ExternalAddrs = NodeAddresses(nodeList, v1.NodeExternalIP)
+	config.ExternalAddrs = e2enode.FirstNodeAddresses(nodeList, v1.NodeExternalIP)
```
b/c you now have a node package the naming of the function signatures should change and should probably be done in this PR. 

e2enode.FirstAddresses()... 

This goes for all other signatures that will now have a redundant "node" in the name

jiatongw replied:
For some of the functions, changes like `WaitForNodeToBe` -> `WaitForToBe`, `CheckNodesReady`-> `CheckReady`, `NumberOfReadyNodes` -> `NumberOfReady` are ambiguous. 
How about remove part of the redundant strings but keep some that described as above?

timothysc reviewed (changes requested):
see other comment

jiatongw commented:
Changed as requested @timothysc 

Renamed exported functions whose name have redundant strings. I leave unexported functions unchanged because those will not be used as `e2enode.xxx`, and it's readable since some of them don't have comments. 

I have a second commit for reviewers convenience. Once approved, I'll squash to one.

jiatongw commented:
/retest

jiatongw commented:
/test pull-kubernetes-e2e-gce

timothysc reviewed (changes requested):
minor comments, then lgtm 
/approve

alejandrox1 commented on test/e2e/framework/node/wait.go:
```diff
@@ -0,0 +1,197 @@
+// WaitForReady waits up to timeout for cluster to has desired size and
+// there is no not-ready nodes in it. By cluster size we mean number of Nodes
+// excluding Master Node.
+func WaitForReady(c clientset.Interface, size int, timeout time.Duration) error {
```
Why change the signature of this method from `e2enode.WaitForReadyNodes` to `e2enode.WaitForReady`?
1WaitForReady` could wait for any general condition while `WaitForReadyNodes` makes it sound more explicit that the condition is a certain number of nodes

jiatongw replied:
As per Tim's suggestion, we have a `e2enode` package, so the signature should remove the redundant `node(s)` in naming. `e2enode.WaitForReady` has a `e2enode` prefix, so people know that this is a node related function

what do you think?

alejandrox1 replied:
Makes sense removing the redundant naming but for this one, if i was looking at godos (as an example), `e2enode.WaitForReadyNodes` sounds like it waits for some number of nodes, while `e2enode.WaitForReady` sounds like i could pass some "condition" object and check if a certain node meets that condition.

jiatongw replied:
👍 make more sense

alejandrox1 commented on test/e2e/framework/node/wait.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+// WaitForNameToBeReady returns whether node name is ready within timeout.
+func WaitForNameToBeReady(c clientset.Interface, name string, timeout time.Duration) bool {
```
Here you change the signature from `WaitForNodeToBeReady` to `WaitForNameToBeReady`, yes?
wouldn't it read better `e2enode.WaitForNodeToBeReady`? 
I think the name makes more sense given what it does.

jiatongw replied:
same as above

alejandrox1 replied:
For this one, `e2enode.WaitForNameToBeReady` sounds a bit weird to me, whereas `e2enode.WaitForNodeToBeReady` tells me exactly what it does. I think keeping the extra "node" in the method signature adds value (is not purely redundant).

jiatongw replied:
will fix, thanks

alejandrox1 commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,316 @@
+)
+
+const (
+	// Poll is how often to Poll pods, nodes and claims.
```
s/Poll/poll

jiatongw replied:
thanks for catching this. will fix

alejandrox1 commented on test/e2e/framework/node/resource.go:
```diff
@@ -0,0 +1,316 @@
+	// Poll is how often to Poll pods, nodes and claims.
+	poll = 2 * time.Second
+
+	// SingleCallTimeout is how long to try single API calls (like 'get' or 'list'). Used to prevent
```
s/SingleCallTimeout/singleCallTimeout

timothysc commented:
Just poke me on the #testing-commons slack channel when you feel it's all ready to go.

timothysc reviewed (approved):
Going to call good enough.  I could totally needle stuff, but it's not your code, it's what was inherited. 

/lgtm

alejandrox1 commented:
/retest

jiatongw commented:
@timstclair  Hi Tim, could you re-review and lgtm? Merging conflicts so I rebased.

alejandrox1 reviewed (approved):
/lgtm
